logstash-core/src/main/java/org/logstash/secret/SecretIdentifier.java

```diff
@@ -0,0 +1,101 @@
+package org.logstash.secret;
+
+import java.util.Locale;
+import java.util.regex.Pattern;
+
+/**
+ * <p>A URN based identifier for a given secret. The URN is constructed as such: {@code urn:logstash:secret:v1:<key>}</p>
+ * <p><em>Note - v1 is the version of the URN (not the key). This allows for passive changes to the URN for anything to the right of v1</em></p>
+ */
+public class SecretIdentifier {
+
+    private final static Pattern colonPattern = Pattern.compile(":");
+    private final static Pattern urnPattern = Pattern.compile("urn:logstash:secret:v1:.*$");
+    private final String VERSION = "v1";
+    private final String key;
+
+    /**
+     * Constructor
+     *
+     * @param key The unique part of the identifier. This is the key to reference the secret, and the key itself should not be sensitive. For example: {@code db.pass}
+     */
+    public SecretIdentifier(String key) {
+        this.key = validateWithTransform(key, "key");
+    }
+
+    /**
+     * Converts an external URN format to a {@link SecretIdentifier} object.
+     *
+     * @param urn The {@link String} formatted identifier obtained originally from {@link SecretIdentifier#toExternalForm()}
+     * @return The {@link SecretIdentifier} object used to identify secrets, null if not valid external form.
+     */
+    public static SecretIdentifier fromExternalForm(String urn) {
+        if (urn == null || !urnPattern.matcher(urn).matches()) {
+            return null;
+        }
+        String[] parts = colonPattern.split(urn, 5);
+        return new SecretIdentifier(validateWithTransform(parts[4], "key"));
+    }
+
+    /**
+     * Minor validation and transformation on input. Converts ":" to "_" to avoid URN conflicts, and downcases the parts
+     *
+     * @param part     The part of the URN to validate
+     * @param partName The name of the part used for logging.
+     * @return The validated and transformed part.
+     */
+    private static String validateWithTransform(String part, String partName) {
+        if (part == null || part.isEmpty()) {
+            throw new IllegalArgumentException(String.format("%s may not be null or empty", partName));
+        }
+        return part.replace(":", "_").toLowerCase(Locale.US);
```

Comment:
Why? Is this because you use `/:/.split(urn, 5)` ?

Feels weird to transform the user input from `foo:bar` to `foo_bar` internally. I worry, in debugging, this will cause confusion. Is it necessary?

Reply:
Agree, this is weird. Will change.